### PR TITLE
[#ICC-205]  Implement new notify endpoint

### PR DESCRIPTION
**Pull request by gquadrati**
This PR depends on https://github.com/pagopa/io-backend/pull/929

#### List of Change

 - Add readers for retrieving a message with its content, a service and a user session status.
 - Add sendNotification function for writing notification to `fn-pushnotification` queue
 - Add handler implementation

#### Motivation and Context

Enable Reminder Notifications

#### How Has This Been Tested?

Unit tests

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: Notify/handler.ts [hunk 1]
```diff
@@ -8,24 +13,190 @@ import { RequiredBodyPayloadMiddleware } from "@pagopa/io-functions-commons/dist
...
+  notification_type: NotificationTypeEnum,
+  fiscal_code,
+  message_id
+): TE.TaskEither<
+  IResponseErrorInternal | IResponseErrorNotFound,
+  NotificationPrinter
+> =>
+  pipe(
+    canSendVerboseNotification(retrieveUserSession, fiscal_code),
+    TE.orElse(_err => TE.of(false)),
```
Comment (AleDore) [inline Notify/handler.ts:108]: Are we sure we want to consider an error with a default behaviour? (i.e: BE not reacheable so we send an anonymous notification)

Comment (gquadrati) [reply Notify/handler.ts:108]: This is an open point, you're right.
Is it better to send them anyway anonymously, or is it better to block all notifications till the error is gone?
I don't have a strong opinion on that.

File: Notify/notification.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        title: notificationTitle
+      }
+    },
+    (notifyMessage: NotifyMessage) =>
+      TE.tryCatch(
+        () =>
+          notificationQueueClient.sendMessage(
+            base64EncodeObject(notifyMessage)
+          ),
+        (error: { readonly message?: string }) =>
```
Comment (AleDore) [inline Notify/notification.ts:53]: Is it necessary? Can we infer this type from sendMessage return type?

Comment (AleDore) [reply Notify/notification.ts:53]: What about this?
```
flow(E.toError, error =>
  Error(
    `Error while sending notify message to the queue [${error.message}]`
  )
)
```

**Review by AleDore — APPROVED** [review]

File: utils/config.ts [hunk 3]
```diff
@@ -81,6 +81,13 @@ export const IConfig = t.intersection([
     isProduction: t.boolean
     /* eslint-enable sort-keys */
   }),
+  t.interface({
```
Comment (AleDore) [inline utils/config.ts:84]: Why creating an anonymous interface instead of fill env variables directly into the previous interface?

Comment (gquadrati) [reply utils/config.ts:84]: No reason, actually, I'll fix it.

Comment (gquadrati) [reply utils/config.ts]: Done in [a80e8ac](https://github.com/pagopa/io-functions-service-messages/pull/14/commits/a80e8ac41a6102b1d343ce02df8b75a9cde279c5)

File: Notify/notification.ts (hunk 2 above)
Comment (gquadrati) [reply Notify/notification.ts]: Done in [d645e45](https://github.com/pagopa/io-functions-service-messages/pull/14/commits/d645e4548ba1a8623eb131cfeefe49696e208085) and added tests

File: Notify/notification.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    {
+      installationId: toHash(fiscalCode) as NonEmptyString,
+      kind: NotifyKind[NotificationMessageKindEnum.Notify],
+      payload: {
+        message: notificationBody,
+        message_id: messageId,
+        title: notificationTitle
+      }
+    },
+    (notifyMessage: NotifyMessage) =>
```
Comment (AleDore) [inline Notify/notification.ts:48]: do we need explicit typings here?

File: Notify/readers.ts [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+): TE.TaskEither<
+  IResponseErrorInternal | IResponseErrorNotFound,
+  RetrievedMessage
+> =>
+  pipe(
+    messageModel.findMessageForRecipient(fiscalCode, messageId),
+    TE.mapLeft<CosmosErrors, IResponseErrorNotFound | IResponseErrorInternal>(
+      _ => ResponseErrorInternal("Error while retrieving the message metadata")
+    ),
+    TE.chainOptionK(() =>
```
[the remark this answers is not in the document]
Comment (gquadrati) [reply Notify/readers.ts]: I was expecting this comment :D Unfortunately we would need a `chainOptionKW`
Done in [8b08f43](https://github.com/pagopa/io-functions-service-messages/pull/14/commits/8b08f43479f253a120006e2aba6ed977eda258be)

File: Notify/notification.ts (hunk 4 above)
Comment (gquadrati) [reply Notify/notification.ts:48]: I'd like to have a static check over `NotifyMessage` type but I didn't want to add a `decode` step

File: Notify/readers.ts [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+export type SessionStatusReader = (
+  fiscalCode: FiscalCode
+) => TE.TaskEither<IResponseErrorInternal, UserSessionInfo>;
+
+export const getUserSessionStatusReader = (
+  sessionClient: SessionClient<"token">
+): SessionStatusReader => (fiscalCode): ReturnType<SessionStatusReader> =>
+  pipe(
+    TE.tryCatch(
+      async () => sessionClient.getSession({ fiscalcode: fiscalCode }),
```
Comment (AleDore) [inline Notify/readers.ts:143]: Do we need `async` here?

Comment (gquadrati) [reply Notify/readers.ts:143]: I thought it could be useful for the abortable fetch, but I didn't try actually

**Review by AleDore — APPROVED** [review]
LGTM

**Review by AleDore — APPROVED** [review]
LGTM

**Pull request merged**